Integrate Android logcat to wptrunner

WeizhongX opened the pull request:
When running test against android targets(clank, webview, weblayer),
pull logcat with command "adb logcat *:D" on the fly, and save
that to result viewer. Logs are only saved for CRASH case for now.

Test: Use geolocation-API/PositionOptions.https.html as an example,
the waterfall run shows the crashlog missing, but it's present in this CL

Bug: 1071136

WeizhongX committed:
Integrate Android logcat to wptrunner

When running test against android targets(clank, webview, weblayer),
pull logcat with command "adb logcat *:D" on the fly, and save
that to result viewer. Logs are only saved for CRASH case for now.

Test: Use geolocation-API/PositionOptions.https.html as an example,
the waterfall run shows the crashlog missing, but it's present in this CL

Bug: 1071136

stephenmcgruer commented:
Close/opening to re-run taskcluster now that https://github.com/web-platform-tests/wpt/pull/27355 has landed.

stephenmcgruer commented:
From tools/ unittests, some style violations to be cleaned up:

```
./wptrunner/wptrunner/browsers/android_webview.py:1:1: F401 'subprocess' imported but unused
./wptrunner/wptrunner/browsers/android_webview.py:3:1: F401 '.base.ExecutorBrowser' imported but unused
./wptrunner/wptrunner/browsers/android_webview.py:6:1: F401 '.chrome_android.AndroidBrowser' imported but unused
./wptrunner/wptrunner/browsers/android_webview.py:7:1: F401 '..webdriver_server.ChromeDriverServer' imported but unused
./wptrunner/wptrunner/browsers/android_webview.py:99:73: W292 no newline at end of file
./wptrunner/wptrunner/browsers/android_weblayer.py:1:1: F401 'subprocess' imported but unused
./wptrunner/wptrunner/browsers/android_weblayer.py:3:1: F401 '.base.Browser' imported but unused
./wptrunner/wptrunner/browsers/android_weblayer.py:3:1: F401 '.base.ExecutorBrowser' imported but unused
./wptrunner/wptrunner/browsers/android_weblayer.py:7:1: F401 '..webdriver_server.ChromeDriverServer' imported but unused
./wptrunner/wptrunner/browsers/android_weblayer.py:101:73: W292 no newline at end of file
./wptrunner/wptrunner/browsers/chrome_android.py:116:9: E722 do not use bare 'except'
```

stephenmcgruer commented on tools/wptrunner/wptrunner/browsers/chrome_android.py:
```diff
@@ -68,21 +69,72 @@ def env_options():
+        except:
+            pass
+
+class AndroidBrowser(Browser):
```
I'm not sure I love the name `AndroidBrowser`, because there are many browsers on Android (Chrome, Firefox, Edge, Brave, ...) and this code doesn't support anything outside of Chrome. 

What is the purpose of the `ChromeAndroidBrowser` subclass? All it does is override `setup_adb_reverse`, which seems to have the same override in all three subclasses. Couldn't we just move `setup_adb_reverse` to the super-class (`AndroidBrowser`), delete the current `ChromeAndroidBrowser` and rename `AndroidBrowser --> ChromeAndroidBrowser`.

stephenmcgruer replied:
Weizhong and I discussed this, and there is a medium bit of refactoring to do to simplify all this further. I don't want to block this PR, so I agreed that we can do the cleanup in a follow-up PR, and for now we will just rename:

`AndroidBrowser --> ChromeAndroidBrowserBase`.

WeizhongX replied:
Hi Stephen
Thanks for the review. I have pushed a new commit. Can you pls review again?
Thanks!

stephenmcgruer reviewed (changes requested):
(Sorry, did not mean to approve this yet -_-)

WeizhongX committed:
Clean up more for chromium android browsers

Rename AndroidBrowser to ChromiumAndroidBrowserBase
Remove unused imports, Correct base class for SystemWebViewShell
Move setup_adb_reverse to base class, fix some exception handling

stephenmcgruer commented on tools/wptrunner/wptrunner/formatters/chromium.py:
```diff
@@ -272,5 +272,6 @@ def suite_end(self, data):
 
     def process_output(self, data):
-        if 'command' in data and 'chromedriver' in data['command']:
+        cmd = data.get('command', "")
```
```suggestion
        cmd = data.get("command", "")
```

Nit: Don't mix `'` and `"` in the same line. (Preferably not in the same file, but I acknowledge that this file already has mismatching uses!).

stephenmcgruer reviewed (approved):
LGTM modulo one nit and the remaining style problems reported by the CI:

```
./wptrunner/wptrunner/browsers/android_webview.py:5:1: F401 '..webdriver_server.ChromeDriverServer' imported but unused
./wptrunner/wptrunner/browsers/android_webview.py:90:1: W391 blank line at end of file
./wptrunner/wptrunner/browsers/android_weblayer.py:5:1: F401 '..webdriver_server.ChromeDriverServer' imported but unused
./wptrunner/wptrunner/browsers/android_weblayer.py:92:1: W391 blank line at end of file
./wptrunner/wptrunner/browsers/chrome_android.py:206:1: W391 blank line at end of file
```

WeizhongX committed:
Fix mix use of quotation mark

WeizhongX committed:
More clean up

stephenmcgruer committed:
Attempt to fix W391
